py/kubeflow/testing/bulk_deploy.py

```diff
@@ -0,0 +1,294 @@
+
+class BulkDeploy:
+  def _read_csv_file(self, projects_path, admin_project, output_dir):
+    bucket, blob_path = util.split_gcs_uri(projects_path)
+
+    client = storage.Client(project=admin_project)
+    bucket = client.get_bucket(bucket)
+
+    blob = bucket.get_blob(blob_path)
+
+    csv_file = os.path.join(output_dir, os.path.basename(blob_path))
+    blob.download_to_filename(csv_file)
+
+    rows = []
+    with open(csv_file) as hf:
+      reader = csv.DictReader(hf)
+      rows = [r for r in reader]
+
+    return rows
+
+  def _load_oauth_file(self, oauth_file, admin_project):
+    bucket, blob_path = util.split_gcs_uri(oauth_file)
+
+    client = storage.Client(project=admin_project)
+    bucket = client.get_bucket(bucket)
+
+    blob = bucket.get_blob(blob_path)
+    contents = blob.download_as_string()
+
+    return yaml.load(contents)
+
+  def _default_job_file(self):
+    this_dir = os.path.dirname(__file__)
+    job_file = os.path.join(this_dir, "..", "..", "..", "codelabs",
+                            "setup-codelab-project.yaml")
+    job_file = os.path.abspath(job_file)
+    return job_file
+
+  def deploy(self, projects_path, job_file=None, output_dir=None,
+             admin_project="kf-codelab-admin", namespace=DEFAULT_NAMESPACE):
+    """Fire off a bunch of K8s jobs to deploy many Kubeflow instances.
+
+    Args:
+      projects_path: The GCS path of a CSV file containing the list of projects
+        to deploy in.
+
+      job_file: The path to the YAML file containing a K8s Job that serves
+        as the template for the jobs to be launched.
+
+      output_dir: Directory to write the job specs to.
+    """
+    util.load_kube_config(persist_config=False)
+    # Create an API client object to talk to the K8s master.
+    api_client = k8s_client.ApiClient()
+    batch_api = k8s_client.BatchV1Api(api_client)
+
+    if not job_file:
+      job_file = self._default_job_file()
+
+    if not os.path.exists(job_file):
+      raise ValueError("job file {0} does not exist".format(job_file))
+
+    logging.info("Job file: %s", job_file)
+
+    if not output_dir:
+      output_dir = tempfile.mkdtemp()
+
+    logging.info("output_dir: %s", output_dir)
+
+    rows = self._read_csv_file(projects_path, admin_project, output_dir)
+
+    # Generate a common label for all the jobs. This way we can potentially
+    # wait for all the jobs based on the label.
+    group_label = (datetime.datetime.now().strftime("%Y%m%d-%H%M%S-") +
+                   uuid.uuid4().hex[0:4])
+    for row in rows:
+      project = None
+      user = None
+      for k, v in row.items():
+        if k.lower() == "project":
+          project = v
+        if k.lower() == "username":
+          user = v
+
+      if not project or not user:
+        logging.info("Skipping row %s; could not get project and user", row)
+        continue
+
+      logging.info("Processing project=%s user=%s", project, user)
+
+      # Load a new copy of the job template on each run
```

Comment:
Can we extract line 118-160 to a func `create_job_yaml(...)`?

Reply:
Done.